File: src/analyzer.py
"""統計分析エンジン"""

import logging
import math
from collections import Counter, defaultdict
from typing import Optional

from src.database import get_all_draws

logger = logging.getLogger(__name__)

NUM_RANGE = range(1, 44)  # 1〜43


def _extract_numbers(draw: dict) -> list[int]:
    """抽選結果から本数字6個をリストで返す"""
    return [draw["num1"], draw["num2"], draw["num3"], draw["num4"], draw["num5"], draw["num6"]]


def get_draws_for_analysis(db_path: str = None, limit: int = None) -> list[dict]:
    """分析用データを取得（古い順）"""
    draws = get_all_draws(db_path, limit)
    draws.reverse()  # 古い順に
    return draws


def frequency_analysis(draws: list[dict]) -> dict[int, dict]:
    """各番号(1-43)の出現回数と出現率"""
    total = len(draws)
    if total == 0:
        return {}

    counter = Counter()
    for draw in draws:
        for n in _extract_numbers(draw):
            counter[n] += 1

    result = {}
    for n in NUM_RANGE:
        count = counter.get(n, 0)
        result[n] = {
            "count": count,
            "rate": count / (total * 6) if total > 0 else 0,
            "appearance_rate": count / total if total > 0 else 0,
        }
    return result


def odd_even_analysis(draws: list[dict]) -> dict[str, int]:
    """奇数:偶数の比率パターン分布"""
    patterns = Counter()
    for draw in draws:
        nums = _extract_numbers(draw)
        odd = sum(1 for n in nums if n % 2 == 1)
        even = 6 - odd
        pattern = f"{odd}:{even}"
        patterns[pattern] += 1
    return dict(patterns.most_common())


def range_distribution(draws: list[dict]) -> dict[str, dict]:
    """番号帯別出現割合"""
    ranges = {
        "1-9": range(1, 10),
        "10-19": range(10, 20),
        "20-29": range(20, 30),
        "30-39": range(30, 40),
        "40-43": range(40, 44),
    }
    total_nums = len(draws) * 6
    if total_nums == 0:
        return {}

    counter = Counter()
    for draw in draws:
        for n in _extract_numbers(draw):
            for label, r in ranges.items():
                if n in r:
                    counter[label] += 1
                    break

    result = {}
    for label in ranges:
        count = counter.get(label, 0)
        result[label] = {
            "count": count,
            "rate": count / total_nums,
        }
    return result


def sum_statistics(draws: list[dict]) -> dict:
    """6個の合計値の統計"""
    if not draws:
        return {}

    sums = [sum(_extract_numbers(d)) for d in draws]
    n = len(sums)
    mean = sum(sums) / n
    sorted_sums = sorted(sums)
    median = sorted_sums[n // 2] if n % 2 == 1 else (sorted_sums[n // 2 - 1] + sorted_sums[n // 2]) / 2
    variance = sum((s - mean) ** 2 for s in sums) / n
    std = math.sqrt(variance)

    q1_idx = n // 4
    q3_idx = 3 * n // 4
    return {
        "mean": round(mean, 2),
        "median": median,
        "std": round(std, 2),
        "min": min(sums),
        "max": max(sums),
        "q1": sorted_sums[q1_idx],
        "q3": sorted_sums[q3_idx],
    }


def consecutive_analysis(draws: list[dict]) -> dict:
    """連番ペアの出現率"""
    total = len(draws)
    if total == 0:
        return {"rate": 0, "count": 0}

    consec_count = 0
    pair_counter = Counter()
    for draw in draws:
        nums = sorted(_extract_numbers(draw))
        has_consec = False
        for i in range(len(nums) - 1):
            if nums[i + 1] - nums[i] == 1:
                has_consec = True
                pair_counter[(nums[i], nums[i + 1])] += 1
        if has_consec:
            consec_count += 1

    return {
        "rate": consec_count / total,
        "count": consec_count,
        "total": total,
        "top_pairs": pair_counter.most_common(10),
    }


def dormancy_analysis(draws: list[dict]) -> dict[int, int]:
    """各番号の出遅れ度（最後に出てからの経過回数）"""
    if not draws:
        return {}

    last_seen = {}
    latest_id = draws[-1]["id"] if draws else 0

    for draw in draws:
        for n in _extract_numbers(draw):
            last_seen[n] = draw["id"]

    result = {}
    for n in NUM_RANGE:
        if n in last_seen:
            result[n] = latest_id - last_seen[n]
        else:
            result[n] = latest_id  # 一度も出ていない
    return result


def correlation_analysis(draws: list[dict], top_n: int = 20) -> list[tuple]:
    """番号ペアの同時出現頻度 TOP N"""
    pair_counter = Counter()
    for draw in draws:
        nums = sorted(_extract_numbers(draw))
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                pair_counter[(nums[i], nums[j])] += 1

    return pair_counter.most_common(top_n)


def weekday_trend(draws: list[dict]) -> dict[str, dict[int, dict]]:
    """曜日別の出現傾向"""
    from datetime import datetime

    weekday_draws = {"月曜": [], "木曜": [], "その他": []}
    weekday_map = {0: "月曜", 3: "木曜"}

    for draw in draws:
        try:
            dt = datetime.strptime(draw["draw_date"], "%Y-%m-%d")
            key = weekday_map.get(dt.weekday(), "その他")
            weekday_draws[key].append(draw)
        except (ValueError, KeyError):
            continue

    result = {}
    for day, day_draws in weekday_draws.items():
        if day_draws:
            result[day] = frequency_analysis(day_draws)
    return result


def moving_average_trend(draws: list[dict], window: int = 20) -> dict[int, dict]:
    """各番号の出現頻度の移動平均と傾き"""
    if len(draws) < window:
        return {}

    recent = draws[-window:]
    all_freq = frequency_analysis(draws)
    recent_freq = frequency_analysis(recent)

    result = {}
    for n in NUM_RANGE:
        all_rate = all_freq.get(n, {}).get("appearance_rate", 0)
        recent_rate = recent_freq.get(n, {}).get("appearance_rate", 0)
        slope = recent_rate - all_rate
        result[n] = {
            "all_rate": round(all_rate, 4),
            "recent_rate": round(recent_rate, 4),
            "slope": round(slope, 4),
            "trending": "up" if slope > 0.02 else ("down" if slope < -0.02 else "stable"),
        }
    return result


def full_analysis(db_path: str = None, period: int = None) -> dict:
    """全分析を実行して結果を返す"""
    draws = get_draws_for_analysis(db_path, period)
    if not draws:
        logger.warning("分析対象のデータがありません")
        return {}

    logger.info("分析開始: %d 件のデータ", len(draws))

    result = {
        "total_draws": len(draws),
        "frequency": frequency_analysis(draws),
        "odd_even": odd_even_analysis(draws),
        "range_dist": range_distribution(draws),
        "sum_stats": sum_statistics(draws),
        "consecutive": consecutive_analysis(draws),
        "dormancy": dormancy_analysis(draws),
        "correlation": correlation_analysis(draws),
        "weekday_trend": weekday_trend(draws),
        "moving_avg": moving_average_trend(draws),
    }

    logger.info("分析完了")
    return result

File: js/analyzer.js
/* ===== Statistical Analysis Engine (v2.0 - Enhanced) ===== */

const Analyzer = (() => {
  const NUM_RANGE = Array.from({length: 43}, (_, i) => i + 1);

  function _slice(data, lastN) {
    if (!lastN || lastN === 'all' || lastN >= data.length) return data;
    return data.slice(0, lastN);
  }

  function calcFrequency(data, lastN) {
    const d = _slice(data, lastN);
    const counts = {};
    NUM_RANGE.forEach(n => counts[n] = 0);
    d.forEach(draw => draw.nums.forEach(n => counts[n]++));
    const total = d.length;
    const result = {};
    NUM_RANGE.forEach(n => {
      result[n] = {
        count: counts[n],
        rate: total > 0 ? counts[n] / total : 0,
      };
    });
    return result;
  }

  function calcOddEven(data) {
    const patterns = {};
    data.forEach(draw => {
      const odd = draw.nums.filter(n => n % 2 === 1).length;
      const key = `${odd}:${6 - odd}`;
      patterns[key] = (patterns[key] || 0) + 1;
    });
    return patterns;
  }

  function calcRangeDistribution(data) {
    const ranges = {'1-9': 0, '10-19': 0, '20-29': 0, '30-39': 0, '40-43': 0};
    let total = 0;
    data.forEach(draw => draw.nums.forEach(n => {
      total++;
      if (n <= 9) ranges['1-9']++;
      else if (n <= 19) ranges['10-19']++;
      else if (n <= 29) ranges['20-29']++;
      else if (n <= 39) ranges['30-39']++;
      else ranges['40-43']++;
    }));
    const result = {};
    for (const [k, v] of Object.entries(ranges)) {
      result[k] = { count: v, rate: total > 0 ? v / total : 0 };
    }
    return result;
  }

  function calcSumStats(data) {
    if (!data.length) return {};
    const sums = data.map(d => d.nums.reduce((a, b) => a + b, 0));
    const sorted = [...sums].sort((a, b) => a - b);
    const n = sorted.length;
    const mean = sums.reduce((a, b) => a + b, 0) / n;
    const median = n % 2 === 1 ? sorted[Math.floor(n/2)] : (sorted[n/2-1] + sorted[n/2]) / 2;
    const variance = sums.reduce((acc, s) => acc + (s - mean) ** 2, 0) / n;
    const std = Math.sqrt(variance);
    return {
      mean: Math.round(mean * 100) / 100,
      median, std: Math.round(std * 100) / 100,
      min: sorted[0], max: sorted[n-1],
      q1: sorted[Math.floor(n/4)],
      q3: sorted[Math.floor(3*n/4)],
      distribution: _buildHistogram(sums),
    };
  }

  function _buildHistogram(values) {
    const min = Math.min(...values);
    const max = Math.max(...values);
    const bucketSize = 10;
    const start = Math.floor(min / bucketSize) * bucketSize;
    const end = Math.ceil((max+1) / bucketSize) * bucketSize;
    const buckets = {};
    for (let i = start; i < end; i += bucketSize) {
      buckets[`${i}-${i+bucketSize-1}`] = 0;
    }
    values.forEach(v => {
      const key = `${Math.floor(v / bucketSize) * bucketSize}-${Math.floor(v / bucketSize) * bucketSize + bucketSize - 1}`;
      if (buckets[key] !== undefined) buckets[key]++;
    });
    return buckets;
  }

  function calcConsecutive(data) {
    let count = 0;
    const pairs = {};
    data.forEach(draw => {
      const nums = [...draw.nums].sort((a,b) => a-b);
      let hasConsec = false;
      for (let i = 0; i < nums.length - 1; i++) {
        if (nums[i+1] - nums[i] === 1) {
          hasConsec = true;
          const key = `${nums[i]}-${nums[i+1]}`;
          pairs[key] = (pairs[key] || 0) + 1;
        }
      }
      if (hasConsec) count++;
    });
    return { count, total: data.length, rate: data.length > 0 ? count / data.length : 0, pairs };
  }

  function calcDormancy(data) {
    if (!data.length) return {};
    const latestId = data[0].id;
    const lastSeen = {};
    for (let i = data.length - 1; i >= 0; i--) {
      data[i].nums.forEach(n => lastSeen[n] = data[i].id);
    }
    const result = {};
    NUM_RANGE.forEach(n => {
      result[n] = lastSeen[n] ? latestId - lastSeen[n] : latestId;
    });
    return result;
  }

  function calcCorrelation(data, topN = 20) {
    const pairs = {};
    data.forEach(draw => {
      const nums = [...draw.nums].sort((a,b) => a-b);
      for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
          const key = `${nums[i]}-${nums[j]}`;
          pairs[key] = (pairs[key] || 0) + 1;
        }
      }
    });
    return Object.entries(pairs)
      .sort((a, b) => b[1] - a[1])
      .slice(0, topN)
      .map(([pair, count]) => ({ pair, count }));
  }

  function calcWeekdayTrend(data) {
    const mon = data.filter(d => new Date(d.date + 'T00:00:00').getDay() === 1);
    const thu = data.filter(d => new Date(d.date + 'T00:00:00').getDay() === 4);
    return {
      monday: { count: mon.length, frequency: calcFrequency(mon) },
      thursday: { count: thu.length, frequency: calcFrequency(thu) },
    };
  }

  function calcMovingAvg(data, window = 20) {
    if (data.length < window) return {};
    const recent = data.slice(0, window);
    const allFreq = calcFrequency(data);
    const recentFreq = calcFrequency(recent);
    const result = {};
    NUM_RANGE.forEach(n => {
      const allRate = allFreq[n].rate;
      const recentRate = recentFreq[n].rate;
      const slope = recentRate - allRate;
      result[n] = {
        allRate: Math.round(allRate * 10000) / 10000,
        recentRate: Math.round(recentRate * 10000) / 10000,
        slope: Math.round(slope * 10000) / 10000,
        trending: slope > 0.02 ? 'up' : (slope < -0.02 ? 'down' : 'stable'),
      };
    });
    return result;
  }

  // NEW: Gap analysis
  function calcGapStats(data) {
    const allGaps = [];
    let consecDraws = 0;
    let tripleConsecDraws = 0;
    data.forEach(draw => {
      const nums = [...draw.nums].sort((a,b) => a-b);
      const gaps = nums.slice(1).map((n,i) => n - nums[i]);
      allGaps.push(...gaps);
      let hasConsec = false, hasTriple = false;
      for (let i = 0; i < gaps.length; i++) {
        if (gaps[i] === 1) {
          hasConsec = true;
          if (i < gaps.length-1 && gaps[i+1] === 1) hasTriple = true;
        }
      }
      if (hasConsec) consecDraws++;
      if (hasTriple) tripleConsecDraws++;
    });
    const gapMean = allGaps.reduce((a,b)=>a+b,0)/allGaps.length;
    const gapStd = Math.sqrt(allGaps.reduce((a,v)=>a+(v-gapMean)**2,0)/allGaps.length);
    return { gapMean, gapStd, consecRate: consecDraws/data.length, tripleConsecRate: tripleConsecDraws/data.length, allGaps };
  }

  // NEW: 5-zone distribution
  function calcZoneDistribution(data) {
    const patterns = {};
    data.forEach(draw => {
      const z = [0,0,0,0,0];
      draw.nums.forEach(n => {
        if (n<=9) z[0]++; else if (n<=18) z[1]++; else if (n<=27) z[2]++;
        else if (n<=36) z[3]++; else z[4]++;
      });
      const key = z.join('-');
      patterns[key] = (patterns[key]||0) + 1;
    });
    return Object.entries(patterns).sort((a,b)=>b[1]-a[1]).slice(0,10);
  }

  // NEW: Last digit distribution
  function calcLastDigitStats(data) {
    const digitCounts = new Array(10).fill(0);
    data.forEach(draw => draw.nums.forEach(n => digitCounts[n%10]++));
    return digitCounts;
  }

  // NEW: Carryover analysis
  function calcCarryoverStats(data) {
    const distribution = new Array(7).fill(0);
    let bonusCount = 0;
    for (let i = 1; i < data.length; i++) {
      const prev = new Set(data[i-1].nums);
      const carry = data[i].nums.filter(n => prev.has(n)).length;
      distribution[carry]++;
      if (data[i].nums.includes(data[i-1].bonus)) bonusCount++;
    }
    return { distribution, bonusAppearRate: data.length > 1 ? bonusCount/(data.length-1) : 0 };
  }

  function fullAnalysis(data, lastN) {
    const d = _slice(data, lastN);
    return {
      totalDraws: d.length,
      frequency: calcFrequency(d),
      oddEven: calcOddEven(d),
      rangeDist: calcRangeDistribution(d),
      sumStats: calcSumStats(d),
      consecutive: calcConsecutive(d),
      dormancy: calcDormancy(d),
      correlation: calcCorrelation(d),
      weekdayTrend: calcWeekdayTrend(d),
      movingAvg: calcMovingAvg(d),
      gapStats: calcGapStats(d),
      zoneDistribution: calcZoneDistribution(d),
      lastDigitStats: calcLastDigitStats(d),
      carryoverStats: calcCarryoverStats(d),
    };
  }

  return {
    calcFrequency, calcOddEven, calcRangeDistribution, calcSumStats,
    calcConsecutive, calcDormancy, calcCorrelation, calcWeekdayTrend,
    calcMovingAvg, calcGapStats, calcZoneDistribution, calcLastDigitStats,
    calcCarryoverStats, fullAnalysis,
  };
})();
